File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
import sys

def solution(n, wires):
    answer = sys.maxsize
    
    # BFS
    def bfs(start):
        nonlocal graph, visited
        answer = 0
        queue = deque([start])
        while queue:
            node = queue.popleft()
            for connected in graph[node]:
                if visited[connected]:
                    continue
                visited[connected] = True
                answer += 1
                queue.append(connected)
        return answer
    
    # 그래프 구성
    graph = {}
    for i in range(1, n+1):
        graph[i] = []
    for node1, node2 in wires:
        graph[node1].append(node2)
        graph[node2].append(node1)
    
    # 완탐
    for node1, node2 in wires:
        # wires에서 하나씩 끊기
        graph[node1].remove(node2)
        graph[node2].remove(node1)
        # BFS/방문기록
        visited = [False] * (n+1)
        visited[node1] = True
        visited[node2] = True
        cnt1 = bfs(node1)
        cnt2 = bfs(node2)
        # 초기화
        graph[node1].append(node2)
        graph[node2].append(node1)
        # 차이 계속 계산
        answer = min(abs(cnt1-cnt2), answer)

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        // 하나씩 순회하며 자를 wire 구하기
        for (int cut = 0; cut < wires.length; cut++) {
            
            // 트리 > List
            List<List<Integer>> tree = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                tree.add(new ArrayList<>());
            }
            
            // 자를 두 노드는 제외하고 트리(List) 구성
            for (int i = 0; i < wires.length; i++) {
                if (i == cut) continue;
                tree.get(wires[i][0]).add(wires[i][1]);
                tree.get(wires[i][1]).add(wires[i][0]);
            }
            
            // 갱신
            int cnt1 = count(tree, wires[cut][0], n);
            int cnt2 = count(tree, wires[cut][1], n);
            answer = Math.min(answer, Math.abs(cnt1 - cnt2));
        }
        
        return answer;
    }
    
    private int count(List<List<Integer>> tree, int start, int n) {
        boolean[] visited = new boolean[n+1];
        Deque<Integer> stack = new ArrayDeque<>();
        
        stack.push(start);
        visited[start] = true;
        int count = 0;
        
        while (!stack.isEmpty()) {
            int cur = stack.pop();
            count++;
            
            for (int next : tree.get(cur)) {
                if (!visited[next]) {
                    visited[next] = true;
                    stack.push(next);
                }
            }
        }
        return count;
    }
}